### PR TITLE
[Fix] `composed_staged` failing with modifiers with dynamically inferred `end_epoch`

### PR DESCRIPTION
**Pull request by dbogunowicz**
Before:
when incrementing the `end_epoch` value in L213 `optim/manager.py`, the modifiers with attribute `end_epoch=-1` either verbosely or silently failed. 
Now:
this PR allows:
- for modifiers in the last stage of the recipe:  to preserve their implicit (dynamic?) `end_epoch==-1` attribution
- for modifiers not in the last stage of the recipe: to swap `end_epoch=-1` for the largest `end_epoch`, so that for every non-last stage of the recipe each modifier has a clear, fully laid out set of attributes.

### CODE REVIEW

File: tests/sparseml/pytorch/optim/test_manager.py [hunk 1]
```diff
@@ -535,3 +535,100 @@ def test_manager_staged_recipe_serialization(staged_recipe):
...
+
+
+@pytest.mark.parametrize(
+    "base_recipe, additional_recipe, expected_composed_recipe",
+    [
+        # Testing modifiers with dynamic `end_epoch` attribute
+        (SIMPLE_RECIPE, SIMPLE_RECIPE, COMPOSED_RECIPE),
+    ],
+)
+def test_composed_staged(base_recipe, additional_recipe, expected_composed_recipe):
```
Comment (bfineran) [inline tests/sparseml/pytorch/optim/test_manager.py:630]: is there not an existing test that already does this?

File: src/sparseml/optim/manager.py [hunk 2]
```diff
@@ -207,9 +207,28 @@ def compose_staged(
...
+            # make sure that for the modifiers in base_stages
+            # with the initial attribute `end_epoch` = -1,
+            # this attribute value is replaced with `base_end_epoch`
+            for base_modifiers in base_stages.values():
+                for base_modifier in base_modifiers:
+                    if (
+                        hasattr(base_modifier, "end_epoch")
+                        and base_modifier.end_epoch == -1
+                    ):
+                        base_modifier._init_end = base_end_epoch
```
Comment (bfineran) [inline src/sparseml/optim/manager.py:220]: where is _init_end used?

File: tests/sparseml/pytorch/optim/test_manager.py (hunk 1 above)
Comment (dbogunowicz) [reply tests/sparseml/pytorch/optim/test_manager.py:630]: True, there is one, but it does not use those "problematic modifiers". I will merge those tests together.

File: src/sparseml/optim/manager.py (hunk 2 above)
Comment (dbogunowicz) [reply src/sparseml/optim/manager.py:220]: Without _init_end modification, we get ValueError at L707 `optim/modifier.py`:
```
        if self._end_comparator is None and self._end_epoch != self._init_end:
            raise ValueError(
                "end_epoch of {} must be equal the init value of {} for {}".format(
                    self._end_epoch, self._init_end, self.__class__.__name__
                )
            )
```

File: tests/sparseml/pytorch/optim/test_manager.py (hunk 1 above)
Comment (bfineran) [reply tests/sparseml/pytorch/optim/test_manager.py:630]: yeah, since we're using pytest parameterize we can just add in the test strings and not need to duplicate code

File: src/sparseml/optim/manager.py (hunk 2 above)
Comment (bfineran) [reply src/sparseml/optim/manager.py:220]: looks good, one thing we need to eventually consider is some modifiers we may want to have automatically end at the end of a stage if indefinite (maybe constant pruning), but some should stick around forever such as quantization. I think we're ok for now, but as this gets more used and we hit corner cases and new modifiers this may come up

Comment (dbogunowicz) [reply src/sparseml/optim/manager.py:220]: Hmm, this is a super interesting corner case. I like the idea of handling them progressively. At the same time, we may now start documenting them and logging them into the backlog, what do you think?

**Review by bfineran — APPROVED** [review]

File: src/sparseml/optim/manager.py (hunk 2 above)
Comment (KSGulin) [reply src/sparseml/optim/manager.py:220]: @bfineran what's the context for the conditional on optim/modifier.py::707? Because it seems it's designed to prevent modifications of the end epoch and I'm wondering if it's an anti pattern to work around the conditional rather than to modify/delete it

Comment (dbogunowicz) [reply src/sparseml/optim/manager.py:220]: @KSGulin very good point

Comment (bfineran) [reply src/sparseml/optim/manager.py:236]: idea here was if something doesn't have an end comparator then the end epoch likely should be -1 or something else that shouldn't be set which is why this check is in place.

**Review by spacemanidol — APPROVED** [review]
